src/Spatial/Euclidean/LIneSegment3D.cs

```diff
@@ -0,0 +1,317 @@
+        }
+
+        /// <inheritdoc/>
+        [Pure]
+        public override string ToString()
+        {
+            return $"StartPoint: {this.StartPoint}, EndPoint: {this.EndPoint}";
+        }
+
+        /// <summary>
+        /// Returns a value to indicate if a pair of line segments are equal
+        /// </summary>
+        /// <param name="other">The line segment to compare against.</param>
+        /// <param name="tolerance">A tolerance (epsilon) to adjust for floating point error</param>
+        /// <returns>True if the line segments are equal; otherwise false</returns>
+        [Pure]
+        public bool Equals(LineSegment3D other, double tolerance)
+        {
+            return this.StartPoint.Equals(other.StartPoint, tolerance) && this.EndPoint.Equals(other.EndPoint, tolerance);
+        }
+
+        /// <inheritdoc/>
+        [Pure]
+        public bool Equals(LineSegment3D other)
+        {
+            return this.StartPoint.Equals(other.StartPoint) && this.EndPoint.Equals(other.EndPoint);
+        }
+
+        /// <inheritdoc />
+        [Pure]
+        public override bool Equals(object obj)
+        {
+            if (obj is null)
+            {
+                return false;
+            }
+
+            return obj is LineSegment3D d && this.Equals(d);
+        }
+
+        /// <inheritdoc />
+        [Pure]
+        public override int GetHashCode()
+        {
+            unchecked
+            {
+                return (this.StartPoint.GetHashCode() * 397) ^ this.EndPoint.GetHashCode();
+            }
+        }
+
+        /// <summary>
+        /// Extends the segment to a infinite line and finds the closest point on that line to the provided point.
+        /// </summary>
+        /// <param name="p">a point</param>
+        /// <returns>A point on the infinite line which extends the segment</returns>
+        [Pure]
+        private Point3D ClosestLinePointTo(Point3D p)
+        {
+            var alongVector = this.StartPoint.VectorTo(p).DotProduct(this.Direction) * this.Direction;
+            return this.StartPoint + alongVector;
+        }
+
+        /// <summary>
+        /// Computes the pair of points which represent the closest distance between this Line3D and another Line3D, with the first
+        /// point being the point on this Line3D, and the second point being the corresponding point on the other Line3D.  If the lines
+        /// intersect the points will be identical, if the lines are parallel the first point will be the start point of this line.
+        /// </summary>
+        /// <param name="other">line to compute the closest points with</param>
+        /// <param name="tolerance">A tolerance (epsilon) to adjust for floating point error</param>
+        /// <returns>A tuple of two points representing the endpoints of the shortest distance between the two lines</returns>
+        [Pure]
+        private Tuple<Point3D, Point3D> ClosestPointsBetweenLines(LineSegment3D other, Angle tolerance)
```

Comment:
Not sure what is the best signature here, we can merge it as is as it is a port of Line3D.

Reply:
if we add a replacement Line2D class in the future which is not bounded by points, then this method would go away in favor of using whatever the public interface of Line would be, hence I opted for just making it private and leaving it as it was